src/components/Admin/Trainer/FirebaseStorageHandler.js

```diff
@@ -0,0 +1,191 @@
+    if (fileNamePrefix == null || fileNamePrefix.length === 0) {
+      fileNamePrefix = DEFAULT_FILE_NAME_PREFIX;
+    }
+
+    this.modelTopologyFileName = fileNamePrefix + DEFAULT_JSON_EXTENSION_NAME;
+    this.weightDataFileName =
+      fileNamePrefix + DEFAULT_WEIGHT_DATA_EXTENSION_NAME;
+    this.modelName = fileNamePrefix;
+    this.classes = classes;
+  }
+
+  async save(modelArtifacts) {
+    const weightsBlob = new Blob([modelArtifacts.weightData], {
+      type: 'application/octet-stream'
+    });
+
+    if (modelArtifacts.modelTopology instanceof ArrayBuffer) {
+      throw new Error(
+        'DownloadTrigger.save() does not support saving model topology ' +
+          'in binary formats yet.'
+      );
+    } else {
+      const weightsManifest = [
+        {
+          paths: ['./' + this.weightDataFileName],
+          weights: modelArtifacts.weightSpecs
+        }
+      ];
+      const modelTopologyAndWeightManifest = {
+        modelTopology: modelArtifacts.modelTopology,
+        weightsManifest,
+        classes: this.classes
+      };
+      const modelTopologyAndWeightManifestBlob = new Blob(
+        [JSON.stringify(modelTopologyAndWeightManifest)],
+        {type: 'application/json'}
+      );
+
+      initFirebase();
+      const store = firebase.storage();
+
+      await store
+        .ref()
+        .child('models/' + this.modelTopologyFileName)
+        .put(modelTopologyAndWeightManifestBlob);
+
+      if (modelArtifacts.weightData != null) {
+        await store
+          .ref()
+          .child('models/' + this.weightDataFileName)
+          .put(weightsBlob);
+      }
+
+      const modelRef = firebase
+        .firestore()
+        .collection('models')
+        .doc(this.modelName);
+
+      const modelRow = await modelRef.get();
+      if (!modelRow.exists) {
+        modelRef.set({deployed: false});
+      }
+
+      return {
+        modelArtifactsInfo: getModelArtifactsInfoForJSON(modelArtifacts)
+      };
+    }
+  }
+
+  async load() {
+    initFirebase();
+    const store = firebase.storage();
+    this.path = await store
+      .ref()
+      .child('models/' + this.modelTopologyFileName)
+      .getDownloadURL();
+
+    const modelConfigRequest = await fetch(this.path);
+    const modelConfig = await modelConfigRequest.json();
+    const modelTopology = modelConfig['modelTopology'];
+    const weightsManifest = modelConfig['weightsManifest'];
+    const classes = modelConfig['classes'];
+
+    // We do not allow both modelTopology and weightsManifest to be missing.
+    if (modelTopology == null && weightsManifest == null) {
+      throw new Error(
+        `The JSON from HTTP path ${this.path} contains neither model ` +
+          `topology or manifest for weights.`
+      );
+    }
+
+    let weightSpecs;
+    let weightData;
+    if (weightsManifest != null) {
+      const weightsManifest = modelConfig['weightsManifest'];
+      weightSpecs = [];
+      for (const entry of weightsManifest) {
```

Comment:
We could use [`_.flatMap`](https://lodash.com/docs/4.17.10#flatMap) to simplify this a bit

```
const weightSpecs = _.flatMap(modelConfig.weightsManifest, entry => entry.weights)

```